deepinv/sampling/diffusion_discrete.py

```diff
@@ -0,0 +1,519 @@
+    alpha : float
+        (Unused, kept for API compatibility.)
+    num_train_timesteps : int
+        Training horizon :math:`T` used to build schedules.
+    num_inference_steps : int
+        Number of sampling steps.
+    model : LatentDiffusion | None
+        Latent model exposing ``forward(x, t, prompt=...)``, ``encode(x)``, and ``decode(z)``.
+    dtype : torch.dtype
+        Computation dtype.
+    device : torch.device
+        Target device.
+    """
+
+    def __init__(
+        self,
+        beta_min: float = 0.00085,
+        beta_max: float = 0.012,
+        num_train_timesteps: int = 1000,
+        num_inference_steps: int = 200,
+        model: LatentDiffusion | None = None,
+        dtype: torch.dtype = torch.float64,
+        device: torch.device | None = None,
+    ) -> None:
+        super().__init__()
+
+        if device is None:
+            device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+        self.beta_min = float(beta_min)
+        self.beta_max = float(beta_max)
+        self.beta_d = self.beta_max - self.beta_min
+        self.num_train_timesteps = int(num_train_timesteps)
+        self.num_inference_steps = int(num_inference_steps)
+        self.model = model
+        self.dtype = dtype
+        self.device = device
+
+        self.alphas_cumprod = self.get_alpha_prod(
+            beta_start=beta_min,
+            beta_end=beta_max,
+            num_train_timesteps=num_train_timesteps,
+        )[4].to(self.device, self.dtype)
+
+    def get_alpha_prod(
+        self,
+        beta_start: float = 0.1 / 1000,
+        beta_end: float = 20 / 1000,
+        num_train_timesteps: int = 1000,
+    ) -> tuple[Tensor, Tensor, Tensor, Tensor, Tensor]:
+        r"""
+        Build alpha-derived sequences used for timestep mappings.
+
+        Returns (in order):
+            ``reduced_alpha_cumprod``,
+            ``sqrt_recip_alphas_cumprod``,
+            ``sqrt_recipm1_alphas_cumprod``,
+            ``sqrt_1m_alphas_cumprod``,
+            ``sqrt_alphas_cumprod``.
+        """
+        betas = torch.linspace(
+            beta_start, beta_end, num_train_timesteps, dtype=torch.float16
+        )
+        alphas = 1.0 - betas
+        alphas_cumprod = np.cumprod(alphas.cpu(), axis=0)  # \bar{α}_t (NumPy by design)
+
+        ab = torch.as_tensor(alphas_cumprod)
+        sqrt_alphas_cumprod = torch.sqrt(ab)
+        sqrt_1m_alphas_cumprod = torch.sqrt(1.0 - ab)
+        reduced_alpha_cumprod = torch.div(sqrt_1m_alphas_cumprod, sqrt_alphas_cumprod)
+        sqrt_recip_alphas_cumprod = torch.sqrt(1.0 / ab)
+        sqrt_recipm1_alphas_cumprod = torch.sqrt(1.0 / ab - 1.0)
+
+        return (
+            reduced_alpha_cumprod,
+            sqrt_recip_alphas_cumprod,
+            sqrt_recipm1_alphas_cumprod,
+            sqrt_1m_alphas_cumprod,
+            sqrt_alphas_cumprod,
+        )
+
+    def forward(
+        self,
+        sample: Tensor,
+        y: Tensor,
+        forward_model: Physics,
```

Comment:
I would use `physics` here as it is used everywhere in the library
```suggestion
        physics: Physics,
```